m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+            } else {
+                transport = new TUdpClient(socketAddresses[0]);
+            }
+
+            transport.open();
+
+            client = new M3.Client(protocolFactory.getProtocol(transport));
+
+            calcProtocol = new TCompactProtocol.Factory().getProtocol(new TCalcTransport());
+            calc = (TCalcTransport) calcProtocol.getTransport();
+
+            freeBytes = calculateFreeBytes(builder.maxPacketSizeBytes, builder.metricTagSet);
+
+            bucketIdTagName = builder.histogramBucketIdName;
+            bucketTagName = builder.histogramBucketName;
+            bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
+
+            metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+            addAndRunProcessor();
+        } catch (TTransportException | SocketException e) {
+            throw new RuntimeException("Exception creating M3Reporter", e);
+        }
+    }
+
+    private int calculateFreeBytes(int maxPacketSizeBytes, Set<MetricTag> commonTags) {
+        metricBatch.setCommonTags(commonTags);
+        metricBatch.setMetrics(new ArrayList<Metric>());
+
+        int size;
+
+        try {
+            metricBatch.write(calcProtocol);
+            size = calc.getSizeAndReset();
+        } catch (TException e) {
+            LOG.warn("Unable to calculate metric batch size. Defaulting to: %s", DEFAULT_METRIC_SIZE);
+            size = DEFAULT_METRIC_SIZE;
+        }
+
+        int numOverheadBytes = EMIT_METRIC_BATCH_OVERHEAD + size;
+
+        freeBytes = maxPacketSizeBytes - numOverheadBytes;
+
+        if (freeBytes <= 0) {
+            throw new IllegalArgumentException("Common tags serialized size exceeds packet size");
+        }
+
+        return freeBytes;
+    }
+
+    private static String getHostName() {
+        try {
+            return InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            LOG.warn("Unable to determine hostname. Defaulting to: %s", DEFAULT_TAG_VALUE);
+            return DEFAULT_TAG_VALUE;
+        }
+    }
+
+    private void addAndRunProcessor() {
+        phaser.register();
+
+        executor.execute(new Processor());
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            metricQueue.put(SizedMetric.FLUSH);
+        } catch (InterruptedException e) {
+            LOG.warn("Interrupted while trying to queue flush sentinel");
+        }
+    }
+
+    private void flush(List<Metric> metrics) {
+        flush(metrics, DEFAULT_FLUSH_TRY_COUNT);
+    }
+
+    private void flush(List<Metric> metrics, int maxTries) {
+        if (metrics.isEmpty()) {
+            return;
+        }
+
+        TException exception = null;
+
+        int tries = 0;
+        boolean tryFlush = true;
+
+        synchronized (metricBatch) {
```

Comment:
Does making metricBatch an instance member instead of alloced/cleaned up in the flush call buy you anything?

Reply:
Good point. Made metricBatch a local variable and add commonTags to it instead - as discussed offline, should help with throughput as this removes the synchronized block.